app/components-react/obs/ObsForm.tsx

```diff
@@ -216,21 +316,272 @@ export function ObsFormGroup(p: IObsFormGroupProps) {
+  const [customResolution, setCustomResolution] = useState(p.inputProps.value);
+
+  // const form = useForm();
+  const formContext = useFormContext();
+
+  function onChange(val: string) {
+    formContext?.antForm.validateFields();
+    // form.validateFields();
+    setCustomResolution(val);
+
+    if (custom && /^[0-9]+x[0-9]+$/.test(customResolution)) {
+      p.inputProps.onChange(customResolution);
+    }
+  }
+
+  function onClick() {
+    formContext?.antForm.validateFields();
+    // form.validateFields();
+    const isValid = /^[0-9]+x[0-9]+$/.test(customResolution);
+    if (custom && !isValid) return;
+
+    if (custom && isValid) {
+      p.inputProps.onChange(customResolution);
+      setCustomResolution('');
+    }
+
+    setCustom(!custom);
+  }
+
+  return (
+    <>
+      {custom ? (
+        <ObsTextInput
+          {...p.inputProps}
+          value={customResolution}
+          onChange={val => onChange(val)}
+          label={p.inputProps.label}
+          validateTrigger="onBlur"
+          rules={[
+            {
+              message: $t('The resolution must be in the format [width]x[height] (i.e. 1920x1080)'),
+              pattern: /^[0-9]+x[0-9]+$/,
+            },
+          ]}
+          uncontrolled={false}
+          name={p.inputProps.name}
+          ref={ref}
+        />
+      ) : (
+        <ListInput {...p.inputProps} allowClear={false} options={p.options} />
+      )}
+
+      <Button
+        type={custom ? 'primary' : 'default'}
+        onClick={onClick}
+        style={{ marginBottom: '24px' }}
+      >
+        {custom ? $t('Apply Primary') : $t('Use Custom')}
+      </Button>
+    </>
+  );
+});
+
+/**
+ * @remark Note: The following is a copy of the shared component `NumberInput` but with the reference forwarded.
+ * This is to allow form validation to work correctly.
+ */
+const ObsNumberInput = forwardRef((p: TNumberInputProps, ref) => {
+  const { inputAttrs, wrapperAttrs, originalOnChange } = useTextInput<typeof p, number>(
+    'number',
+    p,
+    ANT_NUMBER_FEATURES,
+  );
+
+  function onChangeHandler(val: number | string) {
+    // don't emit onChange if the value is out of range
+    if (typeof val !== 'number') return;
+    if (typeof p.max === 'number' && val > p.max) return;
+    if (typeof p.min === 'number' && val < p.min) return;
+    originalOnChange(val);
+  }
+
+  const rules = p.rules ? p.rules[0] : {};
+
+  return (
+    <InputWrapper
+      {...wrapperAttrs}
+      rules={[{ ...rules, type: 'number' }]}
+      style={{ width: '100%' }}
+    >
+      <InputNumber
+        {...inputAttrs}
+        onChange={onChangeHandler}
+        defaultValue={p.defaultValue}
+        ref={ref as RefObject<HTMLInputElement>}
+        style={{ width: '100%' }}
+      />
+    </InputWrapper>
+  );
+});
+
+/**
+ * @remark Note: The following is a copy of the shared component `TextInput` but with the reference forwarded.
+ * This is to allow form validation to work correctly.
+ */
```

Comment:
i wonder if it would make more sense to just forward the refs on these shared components?

Reply:
In order to do that, I have to convert the shared component into a `forwardRef`. Because the `TextInput` and `NumberInput` shared components are used in so many other places, there is a risk that this could effect another form. So in order to do this, we'd have to test every form in the app, which we can do but would be a really large task.